In [1]:
# Libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
# Load data and read
path= '/content/drive/MyDrive/MyProjects/ML Projects/Unsupervised ML/BookRecommendationSystem/data/Books.csv'
df= pd.read_csv(path)
df.head()

<ipython-input-2-db70caa9e810>:3: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  df= pd.read_csv(path)


ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [3]:
# Function displays shape, duplicates count and list of columns
def knowYourData(df):
  print('Shape: ', df.shape)
  print('Duplicates count: ', df.duplicated().sum())
  print('Columns:')
  print(list(df.columns))

knowYourData(df)

Shape:  (271360, 8)
Duplicates count:  0
Columns:
['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher', 'Image-URL-S', 'Image-URL-M', 'Image-URL-L']


In [4]:
# Info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 8 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   ISBN                 271360 non-null  object
 1   Book-Title           271360 non-null  object
 2   Book-Author          271358 non-null  object
 3   Year-Of-Publication  271360 non-null  object
 4   Publisher            271358 non-null  object
 5   Image-URL-S          271360 non-null  object
 6   Image-URL-M          271360 non-null  object
 7   Image-URL-L          271357 non-null  object
dtypes: object(8)
memory usage: 16.6+ MB


In [5]:
# Describe
df.describe()

ISBN      Book-Title      Book-Author  Year-Of-Publication  \
count       271360          271360           271358               271360   
unique      271360          242135           102022                  202   
top     0195153448  Selected Poems  Agatha Christie                 2002   
freq             1              27              632                13903   

        Publisher                                        Image-URL-S  \
count      271358                                             271360   
unique      16807                                             271044   
top     Harlequin  http://images.amazon.com/images/P/185326119X.0...   
freq         7535                                                  2   

                                              Image-URL-M  \
count                                              271360   
unique                                             271044   
top     http://images.amazon.com/images/P/185326119X.0...   
freq                                                    2   

                                              Image-URL-L  
count                                              271357  
unique                                             271041  
top     http://images.amazon.com/images/P/225307649X.0...  
freq                                                    2

In [6]:
# Unique
print(df.nunique())

ISBN                   271360
Book-Title             242135
Book-Author            102022
Year-Of-Publication       202
Publisher               16807
Image-URL-S            271044
Image-URL-M            271044
Image-URL-L            271041
dtype: int64


In [7]:
# Unique value for Year-Of-Publication
print(df['Year-Of-Publication'].unique())

[2002 2001 1991 1999 2000 1993 1996 1988 2004 1998 1994 2003 1997 1983
 1979 1995 1982 1985 1992 1986 1978 1980 1952 1987 1990 1981 1989 1984 0
 1968 1961 1958 1974 1976 1971 1977 1975 1965 1941 1970 1962 1973 1972
 1960 1966 1920 1956 1959 1953 1951 1942 1963 1964 1969 1954 1950 1967
 2005 1957 1940 1937 1955 1946 1936 1930 2011 1925 1948 1943 1947 1945
 1923 2020 1939 1926 1938 2030 1911 1904 1949 1932 1928 1929 1927 1931
 1914 2050 1934 1910 1933 1902 1924 1921 1900 2038 2026 1944 1917 1901
 2010 1908 1906 1935 1806 2021 '2000' '1995' '1999' '2004' '2003' '1990'
 '1994' '1986' '1989' '2002' '1981' '1993' '1983' '1982' '1976' '1991'
 '1977' '1998' '1992' '1996' '0' '1997' '2001' '1974' '1968' '1987' '1984'
 '1988' '1963' '1956' '1970' '1985' '1978' '1973' '1980' '1979' '1975'
 '1969' '1961' '1965' '1939' '1958' '1950' '1953' '1966' '1971' '1959'
 '1972' '1955' '1957' '1945' '1960' '1967' '1932' '1924' '1964' '2012'
 '1911' '1927' '1948' '1962' '2006' '1952' '1940' '1951' '1931' '1954

In [8]:
# Null/ Missing
df.isna().sum()

ISBN                   0
Book-Title             0
Book-Author            2
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

### Data Descriptions:
1. Shape : (271360,8)
2. Duplicate row count : 0
3. Columns : [ 'ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher', 'Image-URL-S', 'Image-URL-M', 'Image-URL-L' ]
4. Types of the columns are Object, later we will transform the data type.
5. Unique counts : ISBN -> 271360, Book-Title -> 242135, Book-Author -> 102022, Year-of-Publication -> 202, Publisher -> 16807.
6. There are some textual and 0 values in year columns.

## Preprocessing

In [9]:
cdf= df.copy() # make copy of the df.
cdf['Year-Of-Publication']= cdf['Year-Of-Publication'].apply(lambda x: str(x)) # Change dtype to str.
def yop_analysis(s):
  if (len(s)==4) & (s.isnumeric()):
    return 1
  elif len(s)>4:
    return 2
  else:
    return 0
cdf['yop']= cdf['Year-Of-Publication'].apply(yop_analysis) # Appying the above function

In [10]:
cdf[cdf['yop']==2]

ISBN                                         Book-Title  \
209538  078946697X  DK Readers: Creating the X-Men, How It All Beg...   
220731  2070426769  Peuple du ciel, suivi de 'Les Bergers\";Jean-M...   
221678  0789466953  DK Readers: Creating the X-Men, How Comic Book...   

       Book-Author Year-Of-Publication  \
209538        2000   DK Publishing Inc   
220731        2003           Gallimard   
221678        2000   DK Publishing Inc   

                                                Publisher  \
209538  http://images.amazon.com/images/P/078946697X.0...   
220731  http://images.amazon.com/images/P/2070426769.0...   
221678  http://images.amazon.com/images/P/0789466953.0...   

                                              Image-URL-S  \
209538  http://images.amazon.com/images/P/078946697X.0...   
220731  http://images.amazon.com/images/P/2070426769.0...   
221678  http://images.amazon.com/images/P/0789466953.0...   

                                              Image-URL-M Image-URL-L  yop  
209538  http://images.amazon.com/images/P/078946697X.0...         NaN    2  
220731  http://images.amazon.com/images/P/2070426769.0...         NaN    2  
221678  http://images.amazon.com/images/P/0789466953.0...         NaN    2

* In above table, values are not in right columns.
* For yop 0 and 1 we will leave them as it is.

In [11]:
# Correcting index 209538
cdf.loc[209538,'Book-Title']= 'DK Readers: Creating the X-Men, How It All Began (Level 4: Proficient Readers)'
cdf.loc[209538,'Book-Author']= 'Michael Teitelbaum'
cdf.loc[209538,'Year-Of-Publication']= '2000'
cdf.loc[209538,'Publisher']= 'DK Publishing Inc'
# # Correcting index 220731
cdf.loc[220731,'Book-Title']= "Peuple du ciel, suivi de 'Les Bergers"
cdf.loc[220731,'Book-Author']= 'Jean-Marie Gustave'
cdf.loc[220731,'Year-Of-Publication']= '2003'
cdf.loc[220731,'Publisher']= 'Gallimard'
# # Correcting index 221678
cdf.loc[221678,'Book-Title']= 'DK Readers: Creating the X-Men, How It All Began (Level 4: Proficient Readers)'
cdf.loc[221678,'Book-Author']= 'James Buckley'
cdf.loc[221678,'Year-Of-Publication']= '2000'
cdf.loc[221678,'Publisher']= 'DK Publishing Inc'

cdf.drop('yop', axis=1, inplace=True)
cdf['Year-Of-Publication']= cdf['Year-Of-Publication'].apply(lambda x: int(x))

In [12]:
# Null/ Missing handling
cols= ['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher'] # List to be kept
cdf['Book-Author'].fillna(cdf['Publisher'], inplace=True)
cdf['Publisher'].fillna(cdf['Book-Author'], inplace=True)
df_result= cdf[cols]
df_result.to_csv('/content/drive/MyDrive/MyProjects/ML Projects/Unsupervised ML/BookRecommendationSystem/data/book_processed.csv')